src/java/org/apache/cassandra/utils/StorageCompatibilityMode.java

```diff
@@ -83,4 +84,25 @@ public void validateSstableFormat(SSTableFormat<?, ?> selectedFormat)
                                                            selectedFormat.name(),
                                                            this));
     }
+
+    /**
+     * Returns the messaging version to use for on-disk storage formats (commit log, hints, batch logs).
+     * When a compatibility mode is set (e.g., CC_4), this ensures that on-disk formats are written
+     * in a way that older versions can read.
+     *
+     * @return the messaging version appropriate for storage serialization
+     */
+    public int storageMessagingVersion()
+    {
+        switch (this)
```

Comment:
nit: we might consider using NoSpamLogger to log that the messaging version is changed by StorageCompatibilityMode.